solutions/gold/kattis-gcdharmony.mdx

```diff
@@ -0,0 +1,196 @@
+
+<Spoiler title="Answer to Hint 2">
+
+If you're only taking values from 2 to 100, the DP won't work-
+it has to account for values *much* higher than that.
+How might we go about doing this without exceeding the time limit?
+
+</Spoiler>
+
+<Spoiler title="Solution">
+
+## Explanation
+
+### Initial DP
+
+Let's keep a DP array (called $dp$) for each node with the definition given in the hints.
+
+$dp$ starts out with the initial value for just changing the node to each starting value.
+For example, if a node's initial value was $5$, then $dp[5]=0$, while $dp[7]=7$.
+
+To flesh out this array, we recursively iterate through all the children and then
+check what their lowest cost value is that is still compatible with each element of the current node's $dp$.
+For example, if we consider $dp[3]$, we would check the *children's* $dp[3]$, $dp[6]$, and so on.
+
+### Refinement
+
+As stated in the hints, just keeping values from 2 to 100 won't work.
+The maximum cost of making a tree harmonious is $2 \cdot 5000=10^4$,
+so we have to keep values all the way up til that.
+
+How do we transition, then, with such large states?
+
+Notice first that it would never make sense for a node to have a final value with a prime factor greater than 100.
+This is true because why would we ever want to change a node to a value with, say, 101 as a prime factor?
+Since a node's initial value can never exceed 100, this extra factor adds additional cost without allowing the node
+to harmonize with any other adjacent nodes.
+
+We can then, for each node, keep track of the minimum cost associated with each prime factor in addition to $dp$.
+For example, the value $7$ in this map would keep track of the minimum cost across $dp[7]$, $dp[14]$, etc.
+We can calculate this map after our $dp$ has been fleshed out, so it won't take too much time.
+
+Then, when fleshing out $dp$, instead of going through all $10^4$ possible child values, we
+look at all the prime factors of the current node value as well as the child's auxiliary map
+to calculate the best cost for the child efficiently.
+
+To see why this saves time, let's say we're calculating $dp[51]$ for a certain node.
+If we stuck with the old transition, we would have to iterate through all $10^4$ values
+for each child and check if they were harmonious with $51$.
+With this new transition, we can get our answer by doing two map lookups for each child:
+one for the $3$ and another for $17$.
+This is far more efficient than the previous approach!
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(n^2 \pi(v))$, where $\pi(v)$ denotes the number of primes less than or equal to $v$.
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <algorithm>
+#include <cstdint>
+#include <iostream>
+#include <map>
+#include <vector>
+
+using std::cout;
+using std::endl;
+using std::vector;
+
+constexpr int MAX_VAL = 100;   // the largest value a node can have
+const int MAX_POSS = 2 * 5e3;  // defined as 2 * (max # of nodes)
```

Comment:
why make the first one constexpr but the second one const?

Reply:
good catch

Reply:
decided to make both const due to precedent